README.md
# rf_distances_filter
Scripts for analyzing Robinson-Foulds distances from Simmons et al. 2016

September 11, 2015
Dan Sloan
dbsloan@rams.colostate.edu

This directory contains three scripts used to compare and organize sets of gene trees based on Robinson-Foulds (RF) distances. There are two Python scripts (RFdistances.py and RFdistances.twofiles.py) that are modified versions of a script written by Sabrina Pankey (https://scriptomika.wordpress.com/2014/01/27/59/). RFdistances.py takes a set of tree files in newick format (one tree per file) as input and reports the RF distance and number of shared taxa for each pairwise comparison between trees. RFdistances.twofiles.py is similar except that the script only takes two tree files as input but each file can contain multiple trees. It returns the RF distance and number of shared taxa for each possible pairwise comparison involving a tree in file 1 and a tree in file 2. There is also a Perl wrapper script (RFdistances.filter.pl) that is used to call RFdistances.py and to return subsets of trees based on the resulting output from RFdistances.py. More details on each script are provided below. The directory also contains a folder of sample input tree files.


General Information and Dependencies

These scripts have been successfully installed on machines running Mac OSX 10.10 with Python 2.7 and Perl 5.18. They have not been tested for broader compatibility.

The python scripts require the following modules:

dendropy (version 3): https://github.com/jeetsukumaran/DendroPy/releases/tag/v3.12.1
p4 :  http://p4.nhm.ac.uk/



RFdistances.py

The RFdistances.py script would typically be called from the command line as follows:

python RFdistances.py -o myOutputFile.txt -p treeFilePrefix

In this case, myOutputFile.txt could be any output file name you choose and treeFilePrefix is a string found at the beginning of all of your desired input tree files (e.g. RAxML_result). The -p argument should include relevant path information if the tree files are not in the present working directory.

The output file contains four tab-delimited columns: 1) first tree; 2) second tree; 3) RF distance between the two trees; 4) number of shared taxa between the two trees.


RFdistances.twoFiles.v2.py

The RFdistances.twoFiles.v2.py script would typically be called from the command line as follows:

python RFdistances.twoFiles.v2.py -a File1 -b File2 -o myOutputFile.txt

File1 and File2 refer to the two tree files (newick format), which can (but do not have to) contain multiple trees each. The output file name can be any that you choose. The output format is the same as for RFdistances.py.



RFdistances.filter.pl


The usage statement for RFdistances.filter.pl is as follows:


Usage: RFdistances.filter.pl [options] treeFilePrefix 

ARGUMENTS
treeFilePrefix - Text shared at the beginning of all input tree files
(Include path if necessary)


OPTIONS
--subsets      - a comma-delimited list of % cutoffs for subet files
default: 10,20,30,40,50,60,70,80,90,100

--output       - a base name for all output files generated
(Include path if necessary)
default: treeFilePrefix

--pyth_script  - full path to RFdistances.py script
default: RFdistances.py

--rand_sets    - number of random subset files to generate                 
default: 0

--RF_file      - full path to file containing output from pairwise RF
calculations to avoid re-running this long step                 
default: 0


EXAMPLE
RFdistances.filter.pl --subsets=80,85,90,95 --output=myOutput RAXML_bestTree


This script calls RFdistances.py, and the treeFilePrefix is the option that will be passed to RFdistances.py with the -p option.

After running RFdistances.py, this script will analyze the output and sort trees based on their average RF distances with the rest of the trees in the dataset. It returns subsets of trees with the lowest average RF distances. By default, it will return subsets with the lowest 10%, 20%, 30%, etc. But the cutoffs for these subsets can also be specified manually with the --subsets option. If the --rand_sets options is specified, the script will also generate random samples of trees with the same set sizes as defined by the --subsets option. 








CHANGES.txt
November 24, 2016

There was a bug in the original version of the "twoFiles" python script. Anytime a pair of trees differs in taxon sampling, the script must prune them to their shared taxa to calculate RF distances. But it was permanently modifying trees when it did this, so when it compared the same tree against a set of others, that tree would quickly be pruned down to almost nothing (assuming taxon sampling was often incomplete), resulting in low numbers of taxa and low RF distances for later comparisons.

This was corrected by using the copy.deepcopy function so that pruning only affected the current pairwise comparison.

The modified version is RFdistances.twoFiles.v2.py



RFdistances.py
#! /usr/bin/env python
import p4, dendropy
import sys, glob
import types
import ast
import exceptions
from optparse import OptionParser
import cStringIO
 
def setuploop(treefilelist,outfile):
    for file in treefilelist:
        genename1 = file #saves last half of filename as gene. should match phytab
        position = treefilelist.index(file)
        allotherfiles = [f for f in treefilelist if treefilelist.index(f) != treefilelist.index(file)]
        for f in allotherfiles:
           genename_f = f
           tree1,tree2,numberOfTaxa = _pruneTreesToComplement(file, f)
 #use dendropy to get distance, save it, and then continue loop
           if tree1 != 'nope':
             t1 = dendropy.Tree.get_from_string(tree1, 'newick')             
             t2 = dendropy.Tree.get_from_string(tree2, 'newick') 
             symdiff = t1.symmetric_difference(t2) #/ len(t1.nodes())
             towrite2 = genename1 + '\t' + genename_f +'\t' + str(symdiff) + '\t' + str(numberOfTaxa) + '\n'
             with open(outfile, 'a') as out2:
                 out2.write(towrite2)
         
def _pruneTreesToComplement(file1,file2):
    p4.var.trees = []
 
    p4.read(file1)
    p4.read(file2)
    tree1 = p4.var.trees[0]
    tree2 = p4.var.trees[1]
   # Delete out the missing taxa from trees so that they intersect:
    tree1_taxa = [n.name for n in tree1.nodes if n.isLeaf]
    tree2_taxa = [n.name for n in tree2.nodes if n.isLeaf]
    commontaxa = list(set(tree1_taxa) & set(tree2_taxa))
    numberOfSharedTaxa = len(commontaxa)
    if len(commontaxa) > 3:
      delete_from_tree1 = set([t for t in tree1_taxa if t not in tree2_taxa])
      delete_from_tree2 = set([t for t in tree2_taxa if t not in tree1_taxa])
      for t in delete_from_tree1:
          tree1.removeNode(t, alsoRemoveBiRoot=False)
      for t in delete_from_tree2:
          tree2.removeNode(t, alsoRemoveBiRoot=False)
      if len([n.name for n in tree1.nodes if n.isLeaf]) != \
            len([n.name for n in tree2.nodes if n.isLeaf]):
          raise TCTError, 'Something unexpected went wrong with the taxon ' + \
                'pruning. After pruning trees have different number of taxa.'
      t1=tree1.writeNewick(toString=True).split('\n')[0]
      t2=tree2.writeNewick(toString=True).split('\n')[0]
      return (t1,t2,numberOfSharedTaxa)
    else:
      return('nope','nope','nope')
def main():
    msg   = 'usage: ./compare_distance.py -p <file prefix> -o <output>' + \
             '\n\nReads two trees (NEXUS and/or PHYLIP format),\n' +\
             'reciprocally prunes each tree of missing taxa ' + \
             'and returns the distance between each pair of trees' + \
             '\n\nDepends on Python2.6, dendopy and p4 modules.\nMust run from within directory containing treefiles.'
    usage = msg
    parser = OptionParser(usage)
    parser.add_option('-p', '--trpref',dest='treeprefix', help='Common prefix shared by all tree files in directory. eg: RAxML_result')
    parser.add_option('-o', '--out',dest='output', help='Name for output containing distances between trees')
    (options,args) = parser.parse_args()
    treeprefix = options.treeprefix
    listoftrees = [i for i in glob.glob(treeprefix+"*")]
    setuploop(listoftrees,options.output)
     
 
if __name__ == "__main__":
    main()

RFdistances.twoFiles.v2.py
#! /usr/bin/env python
import p4, dendropy
import sys, glob
import types
import ast
import exceptions
from optparse import OptionParser
import cStringIO
import copy
 
def setuploop(file_1,file_2,outfile):
	p4.var.trees = []
	p4.read(file_1)
	fileCount1 = len(p4.var.trees)
	p4.read(file_2)
	fileCount2 = len(p4.var.trees) - fileCount1
	
	for i in range(0, fileCount1):
		for j in range(fileCount1, fileCount1+fileCount2):
			
			temp_tree1 = copy.deepcopy(p4.var.trees[i])
			temp_tree2 = copy.deepcopy(p4.var.trees[j])
			
			tree1,tree2,numberOfTaxa = _pruneTreesToComplement(temp_tree1, temp_tree2)
			if tree1 != 'nope':
				t1 = dendropy.Tree.get_from_string(tree1, 'newick')
				t2 = dendropy.Tree.get_from_string(tree2, 'newick')
				symdiff = t1.symmetric_difference(t2) #/ len(t1.nodes())
				towrite2 = file_1 + '\t' + file_2 +'\t' + str(symdiff) + '\t' + str(numberOfTaxa) + '\n'
				with open(outfile, 'a') as out2:
					out2.write(towrite2)
         
def _pruneTreesToComplement(tree1,tree2):
 
   # Delete out the missing taxa from trees so that they intersect:
    tree1_taxa = [n.name for n in tree1.nodes if n.isLeaf]
    tree2_taxa = [n.name for n in tree2.nodes if n.isLeaf]
    commontaxa = list(set(tree1_taxa) & set(tree2_taxa))
    numberOfSharedTaxa = len(commontaxa)
    if len(commontaxa) > 3:
      delete_from_tree1 = set([t for t in tree1_taxa if t not in tree2_taxa])
      delete_from_tree2 = set([t for t in tree2_taxa if t not in tree1_taxa])
      for t in delete_from_tree1:
          tree1.removeNode(t, alsoRemoveBiRoot=False)
      for t in delete_from_tree2:
          tree2.removeNode(t, alsoRemoveBiRoot=False)
      if len([n.name for n in tree1.nodes if n.isLeaf]) != \
            len([n.name for n in tree2.nodes if n.isLeaf]):
          raise TCTError, 'Something unexpected went wrong with the taxon ' + \
                'pruning. After pruning trees have different number of taxa.'
      t1=tree1.writeNewick(toString=True).split('\n')[0]
      t2=tree2.writeNewick(toString=True).split('\n')[0]
      return (t1,t2,numberOfSharedTaxa)
    else:
      return('nope','nope','nope')
def main():
    msg   = 'usage: ./compare_distance.py -a <file 1> -b <file 2> -o <output>' + \
             '\n\nReads two trees (NEXUS and/or PHYLIP format),\n' +\
             'reciprocally prunes each tree of missing taxa ' + \
             'and returns the distance between each pair of trees' + \
             '\n\nDepends on Python2.6, dendopy and p4 modules.\nMust run from within directory containing treefiles.'
    usage = msg
    parser = OptionParser(usage)
    parser.add_option('-a', '--file1',dest='first_file', help='Name of treefile 1')
    parser.add_option('-b', '--file2',dest='second_file', help='Name of treefile 2')
    parser.add_option('-o', '--out',dest='output', help='Name for output containing distances between trees')
    (options,args) = parser.parse_args()
    setuploop(options.first_file,options.second_file,options.output)
     
 
if __name__ == "__main__":
    main()

RFdistances.filter.pl
#!/usr/bin/perl

use strict;
use warnings;
use Scalar::Util qw(looks_like_number);
use List::Util qw(sum shuffle);
use Getopt::Long;
use File::Which;


#parse/define arguments
our $SUBSETS = "10,20,30,40,50,60,70,80,90,100";
our $OUTPUT;
our $PYTHON = which('RFdistances.py');
$PYTHON or $PYTHON = 'RFdistances.py';
our $RAND_SETS = 0;
our $RF_FILE;
parseArgs();
my $prefix = shift;
$OUTPUT or $OUTPUT = $prefix;

#define subsets based on comma-delintied string of percentages
my %subsetContents;
my @splitSubsets = split (/\,/, $SUBSETS);
foreach (@splitSubsets){
	unless (looks_like_number($_) && $_ > 0 && $_ <= 100){
		die ("\nERROR: Specified list of subsets should be comma-delimited and contain non-zero values less than or equal to 100. Offending value: $_\n\n");
	}
	$subsetContents{$_} = "";
}

#check that python script can be found
-e $PYTHON or die ("\nERROR: Could not find the python script $PYTHON\. The full path can be specified with the --pyth_script option.\n\n");

#check that RAND_SETS is a number
looks_like_number($RAND_SETS) or die ("\nERROR: If specified --rand_sets must be followed by a numerical value\n\n");

#call python script unless pairwise RF file has been provided as input
unless ($RF_FILE){ 
	system ("python $PYTHON -p $prefix -o $OUTPUT\.pairwiseRF.txt");
	$RF_FILE = "$OUTPUT\.pairwiseRF.txt";
}

#read in output from python script
my @lines = file_to_array ("$RF_FILE");

#define data structures to store calculated distances and filename info
my %RF_dist_HOA;
my %RF_avgDist;
my @sortedFiles;

#loop over each line in the python script output file
foreach (@lines){
	$_ =~ /^\s*$/ and next;
	chomp $_;
	
	my @sl = split (/\t/, $_);
	
	#calculate corrected RF distance as the raw distance divided by 2 * (number of taxa - 3)
	my $RFcorr = $sl[2]/(2*($sl[3] - 3));
	
	#add corrected RF distance to an array of distances for each tree1 in the input file
	#There will be a separate array for each different tree in column 1, and these are organized in a hash of arrays (tree file names as keys, and refs to distance arrays as values)
	push (@{$RF_dist_HOA{$sl[0]}}, $RFcorr);
}

#Calculate average corrected RF distances for each tree
foreach (keys %RF_dist_HOA){
	my @distArray = @{$RF_dist_HOA{$_}};
	my $avgDist = sum(@distArray)/@distArray;
	$RF_avgDist{$_} = $avgDist;	
}

#sort by avg RF value and print tab-delimited output
my $FH1 = open_output("$OUTPUT\.avg_dist.txt");
foreach (sort { $RF_avgDist{$a} <=> $RF_avgDist{$b} or $a cmp $b } keys %RF_avgDist) {
	print $FH1 "$_\t$RF_avgDist{$_}\n";
	push (@sortedFiles, $_);
}
close $FH1;


#concatenate trees into different subsets based on sorted distances (e.g., the subset 90 includes the 90% of trees with the lowest average RF distances)
my $totalTrees = scalar (keys %RF_avgDist);
my $treeNum = 0;
foreach my $treeFile (@sortedFiles){
	++$treeNum;
	my $fileContents = file_to_string ($treeFile);
	foreach (keys %subsetContents){
		if ($treeNum/$totalTrees <= $_/100){
			$subsetContents{$_} .= $fileContents;
		}
	}	
}

#print concated tree output to files
foreach (keys %subsetContents){
	my $FH2 = open_output ("$OUTPUT\.subset$_\.tre");
	print $FH2 $subsetContents{$_};
}

#generate and print random subsets (as above but with a randomly shuffled list of files)
if ($RAND_SETS){
	my $rand_count = 1;
	while ($rand_count <= $RAND_SETS){
		my @randFiles = shuffle @sortedFiles;
		my $randTreeNum = 0;
		my %randSubsetContents;
		foreach my $randTreeFile (@randFiles){
			++$randTreeNum;
			my $fileContents = file_to_string ($randTreeFile);
			foreach (keys %subsetContents){
				if ($randTreeNum/$totalTrees <= $_/100){
					$randSubsetContents{$_} .= $fileContents;
				}
			}	
		}
		foreach (keys %randSubsetContents){
			my $FH3 = open_output ("$OUTPUT\.rand$rand_count\.subset$_\.tre");
			print $FH3 $randSubsetContents{$_};
		}			
		++$rand_count;
	}	
}


sub parseArgs{

  my $usage = 

"\nUsage: $0 [options] treeFilePrefix 
   
   ARGUMENTS
         treeFilePrefix - Text shared at the beginning of all input tree files
                          (Include path if necessary)
         
   	
   OPTIONS
         --subsets      - a comma-delimited list of % cutoffs for subet files
                          default: 10,20,30,40,50,60,70,80,90,100

         --output       - a base name for all output files generated
                          (Include path if necessary)
                          default: treeFilePrefix

         --pyth_script  - full path to RFdistances.py script
                          default: RFdistances.py
                          
         --rand_sets    - number of random subset files to generate                 
                          default: 0
                          
         --RF_file      - full path to file containing output from pairwise RF
                          calculations to avoid re-running this long step                 
                          default: 0


   EXAMPLE
         $0 --subsets=80,85,90,95 --output=myOutput RAXML_bestTree
\n\n";
  
  my $result = GetOptions
    (
     'subsets=s'  => \$SUBSETS,
     'output=s'  => \$OUTPUT,
     'pyth_script=s'  => \$PYTHON,
     'rand_sets=s'  => \$RAND_SETS,
     'RF_file=s'  => \$RF_FILE,
    );
  
  unless( @ARGV == 1 ){
    print $usage and exit;
  }
  
}



# A subroutine to get data from a file given its filename and store as array
sub file_to_array {
	use strict;
	use warnings;

    my($filename) = @_;

    # Initialize variables
    my @filedata = (  );

    unless( open(GET_FILE_DATA, $filename) ) {
        print STDERR "Cannot open file \"$filename\"\n\n";
        exit;
    }

    @filedata = <GET_FILE_DATA>;

    close GET_FILE_DATA;

    return @filedata;
}

# A subroutine to get data from a file given its filename and store as string
sub file_to_string {
	use strict;
	use warnings;

    my($filename) = @_;

    # Initialize variables
    my $filedata;

    unless( open(GET_FILE_DATA, $filename) ) {
        print STDERR "Cannot open file \"$filename\"\n\n";
        exit;
    }
    
    while (<GET_FILE_DATA>){
    	$filedata .= $_;
    }
    
    close GET_FILE_DATA;

    return $filedata;
}

#A subroutine to open output file and return file handle
sub open_output {
	use strict;
	use warnings;

    my($filename) = @_;
    my $fh_output;

    unless(open($fh_output, ">$filename")) {
        print "Cannot open file $filename\n";
        exit;
    }
    return $fh_output;
}
